api/server/sdk/watcher.go

```diff
@@ -0,0 +1,242 @@
+		return status.Error(codes.DeadlineExceeded,
+			"Deadline is reached, server side func exiting")
+	}
+}
+
+func (s *WatcherServer) startVolumeWatcher(ctx context.Context, done chan bool) error {
+	if s.watchConnections == nil {
+		s.watchConnections = make(map[string][]*watchConnection)
+	}
+
+	// wait for driver to be initialized with an non-empty volume watcher
+	for {
+		if s.volumeServer.driver(ctx) == nil {
+			continue
+		}
+
+		volumeChannel, _ := s.volumeServer.driver(ctx).GetVolumeWatcher(&api.VolumeLocator{}, make(map[string]string))
+		if volumeChannel == nil {
+			continue
+		}
+		s.volumeChannel = volumeChannel
+		goto volumeWatch
+	}
+	// volumeChannel should be waiting for incoming events
+volumeWatch:
+	for {
+		select {
+		case vol := <-s.volumeChannel:
+			s.RLock()
+			for _, client := range s.watchConnections[volumeEventType] {
+				go client.callBack(vol)
+			}
+			s.RUnlock()
+		case <-done:
+			logrus.Infof("exiting volume watcher\n")
+			break volumeWatch
+		}
+	}
+
+	return nil
+}
+
+func (w *WatcherServer) volumeWatch(
+	req *api.SdkVolumeWatchRequest,
+	stream api.OpenStorageWatch_WatchServer,
+) error {
+	ctx := stream.Context()
+	if w.volumeServer.cluster() == nil || w.volumeServer.driver(ctx) == nil {
+		return status.Error(codes.Unavailable, "Resource has not been initialized")
+	}
+
+	// if input has deadline, ensure graceful exit within that deadline.
+	deadline, ok := ctx.Deadline()
+	var cancel context.CancelFunc
+	if ok {
+		// create a new context that will get done on deadline
+		ctx, cancel = context.WithTimeout(ctx, time.Until(deadline))
+		defer cancel()
+	}
+
+	client := watchConnection{
+		name:         uuid.New(),
+		eventType:    volumeEventType,
+		eventChannel: make(chan interface{}, 2),
+	}
+
+	w.registerWatcher(&client, volumeEventType)
+	defer w.removeWatcher(client.name, volumeEventType)
+
+	group, _ := errgroup.WithContext(ctx)
+	errChan := make(chan error)
+
+	// spawn err-group process.
+	group.Go(func() error {
+		if vols, err := w.volumeServer.driver(ctx).Enumerate(&api.VolumeLocator{}, nil); err != nil {
+			return status.Errorf(
+				codes.Internal,
+				"Failed to enumerate volumes: %v",
+				err.Error())
+		} else {
+			for _, vol := range vols {
+				if !vol.IsPermitted(ctx, api.Ownership_Read) {
+					continue
+				}
+				resp := convertApiVolumeToSdkReponse(vol)
+				if err := stream.Send(resp); err != nil {
+					return err
+				}
+			}
+
+			for event := range client.eventChannel {
+				var vol *api.Volume
+				if vol, ok = event.(*api.Volume); !ok {
+					continue
+				}
+				if !vol.IsPermitted(ctx, api.Ownership_Read) {
+					continue
+				}
+				resp := convertApiVolumeToSdkReponse(vol)
+				if err := stream.Send(resp); err != nil {
+					return err
+				}
+			}
+
+		}
+		return nil
+	})
+
+	// wait for err-group processes to be done
+	go func() {
+		errChan <- group.Wait()
+	}()
+
+	// wait only as long as context deadline allows
+	select {
+	case err := <-errChan:
```

Comment:
Why not just do this here?
```suggestion
	case err := <-group.Wait():
```

Reply:
we could but we will need a for loop to wait for it. group.Wait() returns an error instead of a channel.